utils/fs-cache.js

```
@@ -0,0 +1,111 @@
+    if (autosaveInterval > 0) {
+      setTimeout(this.dump.bind(this), autosaveInterval);
+    }
+  }
+
+  /**
+   * @param {string} key
+   * @param {ValueType} value
+   */
+  set(key, value) {
+    this.#cache.set(key, value);
+    this.#saveQueue.push(key);
+  }
+
+  /**
+   * @param {string} key
+   */
+  async get(key) {
+    const fromMemory = this.#cache.get(key);
+    if (fromMemory) {
+      return fromMemory;
+    }
+    this.#deleteQueue.push(key);
+
+    try {
+      /** @type {FsCacheEntry} */
+      const { time, value } = await readJSON(this.getFilePath(key));
+      this.#cache._set(key, time, value);
+      return value;
+    } catch {
+      return;
+    }
+  }
+
+  async dump() {
+    await mkdir(this.#dirname, { recursive: true });
+    await this.invalidate();
+
+    const promisesToWriteFile = [];
+    for (const key of this.#saveQueue) {
+      if (this.#cache.expires(key) > 0) {
+        const file = this.getFilePath(key);
+        const entry = this.#cache._get(key);
+        promisesToWriteFile.push(writeJSON(file, { key, ...entry }));
+      }
+    }
+    this.#saveQueue.length = 0;
+    await Promise.all(promisesToWriteFile);
+    return promisesToWriteFile.length;
+  }
+
+  async load() {
+    await mkdir(this.#dirname, { recursive: true });
+
+    const files = await readDir(this.#dirname);
+    /** @type {FsCacheEntry[]} */
+    for (const { key, time, value } of await Promise.all(files.map(readJSON))) {
+      this.#cache._set(key, time, value);
+    }
+  }
+
+  /**
+   * Clears stale entries from in-memory cache as well as disk.
+   */
+  async invalidate() {
+    const invalidatedKeys = this.#cache.invalidate();
+    const keysToDelete = [
+      ...new Set([...invalidatedKeys, ...this.#deleteQueue]),
+    ];
+    this.#deleteQueue.length = 0;
+    const filesToDelete = keysToDelete.map(key => this.getFilePath(key));
+    await Promise.all(filesToDelete.map(file => unlink(file).catch(() => {})));
```

same as above... seems like hiding errors might not be good.